File: dev.py
import requests, json, base64

def access_token(client_id, client_secret, verify = False):
    client_credentials_raw = client_id + ":" + client_secret
    client_credentials_encoded = base64.b64encode(client_credentials_raw.encode("utf-8"))
    client_credentials = client_credentials_encoded.decode("utf-8")
    response = requests.post(
            url = "https://accounts.spotify.com/api/token", 
            data = {"grant_type": "client_credentials"}, 
            headers = {"Authorization": "Basic " + client_credentials},
            verify = verify
        )
    try:
        return response.json()['access_token']
    except:
        print(response.status_code + " - " + response.reason)


token = access_token(r.auth['SPOTIFY_CLIENTID'], r.auth['SPOTIFY_CLIENTSECRET'])

artist_id = '2CiRpkhwIUTCnAThfOxnZW'

request = requests.get(
      url = "https://api.spotify.com/v1/artists/" + artist_id + "/top-tracks", 
      headers = {"Authorization": "Bearer " + access_token(r.auth['SPOTIFY_CLIENTID'], r.auth['SPOTIFY_CLIENTSECRET'])}, 
      params = {"country": "US"},
      verify = False
)

top_tracks = request.json()


request = requests.get(
      url = "https://api.spotify.com/v1/browse/featured-playlists", 
      headers = {"Authorization": "Bearer " + access_token(r.auth['SPOTIFY_CLIENTID'], r.auth['SPOTIFY_CLIENTSECRET'])}, 
      verify = False
)


featured_playlists = request.json()

from pprint import pprint


pprint(featured_playlists)

for playlist in featured_playlists['playlists']['items']:
      playlist['name']
      playlist['id']
      


playlist_id = featured_playlists['playlists']['items'][0]['id']

request = requests.get(
      url = 'https://api.spotify.com/v1/playlists/' + playlist_id  + '/tracks', 
      headers = {"Authorization": "Bearer " + token}, 
      verify = False
)

tracks = request.json()

for artist in tracks['items'][0]['track']['artists']:
      artist['name']


for track in tracks['items']:
      track_name = track['track']['name']
      track_id = track['track']['id']
      track_artists = []
      for artist in track['track']['artists']:
            track_artists.append(artist['name'])


# https://developer.spotify.com/documentation/web-api/reference/#object-audiofeaturesobject

request = requests.get(
      url = 'https://api.spotify.com/v1/audio-features/' + track_id, 
      headers = {"Authorization": "Bearer " + token}, 
      verify = False
)

track_features = request.json()


track_features




# Get token for authentication - pass result into other API calls
def get_access_token(client_id, client_secret, verify = False):
      
      # Encode raw credentials to API spec (Basic)
      client_credentials_raw = str(client_id) + ":" + str(client_secret)
      client_credentials_encoded = base64.b64encode(client_credentials_raw.encode("utf-8"))
      client_credentials = client_credentials_encoded.decode("utf-8")
      
      response = requests.post(
            url = "https://accounts.spotify.com/api/token", 
            data = {"grant_type": "client_credentials"}, 
            headers = {"Authorization": "Basic " + client_credentials},
            verify = verify
      )
      
      try:
            token = response.json()['access_token']
            return token
      except:
            print(response.status_code + " - " + response.reason)



# Browse featured playlists
def get_featured_playlists(token, verify = True):
      response = requests.get(
            url = "https://api.spotify.com/v1/browse/featured-playlists", 
            headers = {"Authorization": "Bearer " + token}, 
            verify = verify
      )
      try:
            featured_playlists = response.json()
            return featured_playlists
      except:
            print(response.status_code + " - " + response.reason)
      
      



# Get playlist items using playlist ID
def get_playlist(playlist_id, token, verify = True):
      response = requests.get(
            url = 'https://api.spotify.com/v1/playlists/' + str(playlist_id)  + '/tracks', 
            headers = {"Authorization": "Bearer " + token}, 
            verify = verify
      )
      try:
            tracks = response.json()
            return tracks
      except:
            print(response.status_code + " - " + response.reason)



# Get track features using track ID - https://developer.spotify.com/documentation/web-api/reference/#object-audiofeaturesobject
def get_track_features(track_id, token, verify = True):
      response = requests.get(
            url = 'https://api.spotify.com/v1/audio-features/' + str(track_id), 
            headers = {"Authorization": "Bearer " + token}, 
            verify = verify
      )
      try:
             track_features = response.json()
             return track_features
      except:
            print(response.status_code + " - " + response.reason)
     






request = requests.get(
      url = 'https://api.spoonacular.com/recipes/random?apiKey=' + r.auth['SPOONACULAR_APIKEY'] + '&number=100',
      verify = False
)

recipes = request.json()

len(recipes['recipes'])

from pprint import pprint

pprint(recipes['recipes'][0])



request = requests.get(
      url = 'https://api.spoonacular.com/recipes/random?apiKey=' + r.auth['SPOONACULAR_APIKEY'] + '&number=2',
      verify = False
)

recipes = request.json()



for recipe in recipes['recipes']:
      recipe['title']


# Get random recipes
def get_random_recipes(n, apiKey, verify = True):
      request = requests.get(
            url = 'https://api.spoonacular.com/recipes/random?apiKey=' + str(apiKey) + '&number=' + str(n),
            verify = verify
      )


with open('data/recipes_random.json', 'w') as file:
      json.dump(recipes, file)

File: dev.R
library(tidyverse)
library(httr)
library(reticulate)
library(base64enc)
library(jsonlite)


auth <- list(
      SPOTIFY_CLIENTID = Sys.getenv("SPOTIFY_CLIENTID"),
      SPOTIFY_CLIENTSECRET = Sys.getenv("SPOTIFY_CLIENTSECRET"),
      SPOONACULAR_APIKEY = Sys.getenv("SPOONACULAR_APIKEY")
)

source_python("spotify.py")




token <- get_access_token(auth$SPOTIFY_CLIENTID, auth$SPOTIFY_CLIENTSECRET, verify = FALSE)

featured_playlists <- get_featured_playlists(token, verify = FALSE)

playlist_ids <- featured_playlists$playlists$items %>%
      map_chr(function(playlist) {
            playlist$id
      })





# Split vector of ids into a list of smaller vectors with a maximum size for API call limits
split_ids_to_chunks <- function(ids, chunk_size = 100) {
      
      # Calculate how many chunks are needed for the unique ids and assign them to a chunk
      ids <- unique(ids)
      chunk_ids <- ceiling(1:length(ids) / chunk_size) 
      
      # List ids by their assigned chunk
      chunk_ids %>%
            unique() %>%
            map(function(chunk) {
                  ids[chunk_ids == chunk]
            })
      
      
}




df_track = featured_playlists$playlists$items %>%
# df_track = featured_playlists$playlists$items[1:2] %>%
      map(function(featured_playlist) {
            
            # Get playlist items
            message("getting playlist: ", featured_playlist$name)
            playlist <- get_playlist(featured_playlist$id, token, verify = FALSE)
            
            
            # Extract details for each track in playlist
            track_details <- playlist$items %>%
                  map(function(track) {
                        
                        # Return a row for each artist
                        artists <- track$track$artists %>%
                              map(function(artist) {
                                    
                                    tibble(
                                          artist_id = artist$id,
                                          artist_name = artist$name,
                                          artist_url = artist$external_urls$spotify
                                    )
                                    
                              }) %>%
                              bind_rows()
                        
                        # Track data
                        tibble(
                              track_id = track$track$id,
                              track_name = track$track$name,
                              track_url = track$track$external_urls$spotify,
                              track_popularity = track$track$popularity,
                              album_id = track$track$album$id,
                              album_name = track$track$album$name,
                              album_url = track$track$album$external_urls$spotify,
                              release_date = track$track$album$release_date,
                              artists
                        )
                        
                  }) %>%
                  bind_rows()
            
            
            # Set up for API calls to Tracks API (limit = 100)
            track_ids <- split_ids_to_chunks(track_details$track_id)
            
            # Get audio features for each track
            message("    getting audio features")
            track_features <- track_ids %>%
                  map(function(chunk) {
                        track_features <- get_track_features(chunk, token, verify = FALSE)
                        track_features$audio_features
                  }) %>%
                  bind_rows()
            
            
            # Set up for API calls to Artists API (limit = 50)
            artist_ids <- split_ids_to_chunks(track_details$artist_id, chunk_size = 50)
            
            # Get genres from artist data
            message("    getting genres")
            genres <- artist_ids %>%
                  map(function(chunk) {
                        
                        artist_details <- get_artists(chunk, token, verify = FALSE)
                        
                        # Get genre data for each artist
                        artist_details$artists %>%
                              map(function(artist) {
                                    
                                    genres <- artist$genres
                                    
                                    # Artists with no genre data return an empty list - set to NA to create dataframe
                                    if (length(genres) < 1) {
                                          genres <- NA
                                    }
                                    
                                    # Return a row for each genre
                                    tibble(
                                          id = artist$id,
                                          genre = genres
                                    )
                                    
                              }) %>%
                              bind_rows()
                        
                  }) %>%
                  bind_rows()
            
            
            # Combine data
            track_details %>%
                  left_join(track_features, by = c("track_id" = "id")) %>%
                  left_join(genres, by = c("artist_id" = "id"))
            
      }) %>%
      bind_rows()

write_rds(x = df_track, file = "data/tracks_featured.rds")





playlist <- get_playlist(featured_playlists$playlists$items[[3]]$id, token, verify = FALSE)

playlist$items[[1]]$track$artists %>%
      map(function(artist) {
            
            tibble(
                  id = artist$id,
                  name = artist$name,
                  url = artist$external_urls$spotify
            )
            
      })





track_ids <- playlist$items %>%
      map_chr(function(track) {
            track$track$id
      })


track_features <- get_track_features(track_ids, token, verify = FALSE)

track_features$audio_features %>% bind_rows()

tracks <- get_tracks(track_ids[1:50], token, verify = FALSE)


tracks$tracks[[1]]


album_ids <- playlist$items %>%
      map_chr(function(track) {
            track$track$album$id
      })

albums <- get_albums(album_ids[1:20], token, FALSE)


albums$albums[[1]]

albums$albums[[1]]$genres

albums$albums %>% map(function(album) names(album))


playlist$items[[1]]$track$artists

artist_ids <- playlist$items %>%
      map(function(track){
            artists <- track$track$artists
            artists %>%
                  map_chr(function(artist) {
                        artist$id
                  })
      }) %>%
      reduce(c)


artists <- get_artists(artist_ids[1:50], token, FALSE)

artists$artists[[1]]$genres

genres <- artists$artists %>%
      map(function(artist) {
            artist$genres
      })



























